File: algorithmAndDataStructure/cpp/bestTimetoBuyandSellStock/bestTimetoBuyandSellStock.cpp
#include <iostream>
#include <limits.h>
#include <vector>

using namespace std;

class Solution {
public:
    int maxProfit(vector<int>& prices) {
        int minBuy= INT_MAX, maxProfit= 0;
        int pricesLen = prices.size();
        for (int i=0; i < pricesLen; i++) {
            int price = prices[i];
            maxProfit = price>minBuy?max(maxProfit, price-minBuy):maxProfit;
            minBuy = min(minBuy, price);
        }
    }
};

File: algorithmAndDataStructure/golang/bestTimetoBuyandSellStock/bestTimetoBuyandSellStock.go
package bestTimetoBuyandSellStock

import (
	"math"
)

func max(x, y int) int {
	if x > y {
		return x
	}
	return y
}

func min(x, y int) int {
	if x > y {
		return y
	}
	return x
}

func maxProfit(prices []int) int {
	minBuy, maxProfit := math.MaxInt32, 0
	for _, price := range prices {
		if price > minBuy {
			maxProfit = max(maxProfit, price-minBuy)
		}
		minBuy = min(minBuy, price)
	}
	return maxProfit
}
